src/api/ContextMetadata.ts

```diff
@@ -0,0 +1,15 @@
+/**
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright FINOS FDC3 contributors - see NOTICE file
+ */
+
+import { AppMetadata } from './AppMetadata';
+
+/**
+ * Metadata relating to a context or intent and context received through the
+ * `addContextListener` and `addIntentListener` functions.
+ */
+export interface ContextMetadata {
+  /** Metadata relating to the app that sent the context and/or intent. */
+  readonly sourceAppMetadata: AppMetadata;
```

Comment:
Oh, good point!

Reply:
I'd need to combine the two PRs (not really a problem). However, I don't think it should be reduced. I've got `AppIdentifier` filed under types (which are input to functions) rather than metadata (output) and everything `AppMetadata` adds to `AppIdentifier` is optional. Hence, its not increasing requirements for data output, but is keep us consistent (every return is an `AppMetadata`) and typing all the possible returned fields.

Further, when implementing this change, I would likely have a cached set of the current `AppMetadata` objects for each instance of an app allowing a quick return or the richer metadata (which is otherwise hard to retrieve as we've got no applications call - you'd probably end up using `findInstances()` or any other function that returns AppMetadata).

Reply:
I agree with @bingenito . Having AppIdentifier being the return (instanceId) would be a much simpler solution.  An instanceId can always be used to query for more metadata if needed.  It would be great to understand the use cases for an app participating in FDC3 to make decisions based on version, description, or other metadata of an app that has sent it a context or intent.  If apps are relying on this kind of metadata to build interactions, then the robustness of the interop seems to be at risk.

Reply:
I don't see how it would be _much_ simpler as `AppMetadata` is also an `AppIdentifier` but with additional optional fields that the desktop agent can return. Granted to have to test if they are present before using them - perhaps we should consider making more of them required as we can now use AppIdentifier for function arguments. 

> An instanceId can always be used to query for more metadata if needed. 

The problem is, it can't. There is no API call to retrieve metadata on `appId`, `instanceId` or `AppIdentifier` directly. You could use `findInstances` on the `appID I suppose, although that relies on at least one instance still existing or it'll return an empty array. If the appId is fully-qualified you could try to resolve the URL to the app directory and query that, but it's complex and there's no guarantee that the appD is running at the expected URL path. The group has previously pushed back on a call to retrieve metadata for all apps via an `applications()` call. Perhaps this is a shortcoming we should deal with by introducing a `getAppMetadata(AppIdentifier)` call instead? 

> It would be great to understand the use cases for an app participating in FDC3 to make decisions based on version, description, or other metadata of an app that has sent it a context or intent. 

It's not needed to make decisions on interop or that would be in the proposal as a use case and would be required in the type returned rather than optional. However, the `title`, `icons`, `description` and  `tooltip` maybe useful for use in a UI and are troublesome to retrieve by other means. That a slim use case I grant you - but if we reduce to that I think we need to make the `AppMetadata` retreivable independently.